### PR TITLE
Feature ai assistant

### PR DESCRIPTION
**Pull request by PABixn**

### CODE REVIEW

File: Engine-Editor/src/EditorActions.cpp [hunk 1]
```diff
@@ -0,0 +1,1483 @@
...
+		{
+			if (params.size() < 1)
+				return "No entity name provided.";
+
+			Entity entity = m_Scene->FindEntityByName(params.at(0));
+
+			if (entity == Entity())
+				return "Entity with name " + params.at(0) + " not found";
+
+			return std::to_string(entity.GetUUID());
```
Comment (szymix200006) [inline Engine-Editor/src/EditorActions.cpp]: Bring out the actions the functions.

File: Engine-Editor/src/EditorLayer.cpp [hunk 2]
```diff
@@ -630,11 +645,13 @@ namespace eg
 			if (!ImGuizmo::IsUsing())
 			{
 				if (!ImGuizmo::IsUsing())
-					m_GizmoType = -1;
+				{
+					if (!ImGuizmo::IsUsing())
+						m_GizmoType = -1;
+					break;
```
Comment (szymix200006) [inline Engine-Editor/src/EditorLayer.cpp]: What  happened here?

File: Engine-Editor/src/Panels/AssistantPanel.cpp [hunk 3]
```diff
@@ -0,0 +1,597 @@
...
+				"- Muniffic.docx - general documentation of the engine, explaining how the whole engine works.\n"
+				"- Scripting-Engine.cs - scripting engine code containing all functions that can be used for scripting.\n"
+				"- Scripting-Engine-Examples.cs - general rules for writing scripts and examples of how to perform specific actions (how to use scripting system functions).\n"
+				"You answer questions only about Muniffic, don't write anything about other engines.\n"
+				"If you don't know something or are unsure, make an elegant excuse.\n"
+				"If user asks something about scripts, write an example full script file performing requested action.";
+
+				assistantManager->CreateAssistant("Bob", instructions);
+		}
+		else
```
Comment (szymix200006) [inline Engine-Editor/src/Panels/AssistantPanel.cpp]: Move instructions to a textfile

File: Engine-Editor/src/Panels/AssistantPanel.cpp [hunk 4]
```diff
@@ -0,0 +1,597 @@
...
+		drawList->ChannelsSplit(3);
+		drawList->ChannelsSetCurrent(2);
+
+		float padding = 10.0f;
+		float maxBubbleWidth = ImGui::GetWindowSize().x * 0.6f;
+		float codeToolbarHeight = 30.0f;
+		int buttonIndex = 1;
+
+		std::stringstream stream(message);
+		std::string line, language, msg = "";
```
Comment (szymix200006) [inline Engine-Editor/src/Panels/AssistantPanel.cpp]: Why only msg is set to an empty string

File: Engine-Editor/src/Panels/AssistantPanel.cpp [hunk 5]
```diff
@@ -0,0 +1,597 @@
...
+		{
+			ImGui::Text("Assistant initialization failed");
+			ImGui::End();
+			return;
+		}
+
+		if (assistantManager->IsNewVoiceMessageAvailable())
+		{
+			std::string message = assistantManager->GetLastVoiceMessage();
+			memcpy(buffer, message.data(), message.size());
```
Comment (szymix200006) [inline Engine-Editor/src/Panels/AssistantPanel.cpp]: memory access vilolation possible

File: Engine-Editor/src/Panels/AssistantPanel.cpp [hunk 6]
```diff
@@ -0,0 +1,597 @@
...
+				ImGui::CloseCurrentPopup();
+			}
+
+			ImGui::PopStyleColor(13);
+			ImGui::PopStyleVar(6);
+
+			ImGui::EndPopup();
+		}
+
+		ImGui::BeginChild("Messages", ImVec2(ImGui::GetWindowSize().x, availableHeight), false, ImGuiWindowFlags_HorizontalScrollbar);
```
Comment (szymix200006) [inline Engine-Editor/src/Panels/AssistantPanel.cpp:360]: Move message rendering to seperate function

File: Engine-Editor/src/Panels/AssistantPanel.cpp [hunk 7]
```diff
@@ -0,0 +1,597 @@
...
+			m_showMessageTooltip = false;
+
+		drawList->ChannelsMerge();
+
+		ImGui::EndGroup();
+
+		ImGui::NewLine();
+	}
+
+	void AssistantPanel::OnImGuiRender()
```
Comment (szymix200006) [inline Engine-Editor/src/Panels/AssistantPanel.cpp:315]: Move big parts of a window to a separate functions

File: Engine-Editor/src/Panels/AssistantPanel.cpp [hunk 8]
```diff
@@ -0,0 +1,597 @@
...
+		if (m_isListening && !assistantManager->GetVoiceAssistantListening())
+		{
+			assistantManager->SetVoiceAssistantListening(true);
+			std::thread(&AssistantManager::StartListening, assistantManager.get()).detach();
+		}
+		else if (!m_isListening && assistantManager->GetVoiceAssistantListening())
+		{
+			assistantManager->SetVoiceAssistantListening(false);
+			std::thread(&AssistantManager::StopListening, assistantManager.get()).detach();
+		}
```
Comment (szymix200006) [inline Engine-Editor/src/Panels/AssistantPanel.cpp]: Seperate logic from UI

File: Engine-Editor/src/Panels/ContentBrowserPanel.cpp [hunk 9]
```diff
@@ -268,11 +268,11 @@ namespace eg
 
 		float size = thumbnailSize, offset = padding;
 
-		if (ImGui::SliderFloat("Thumbnail Size", &thumbnailSize, 16, 512))
+		/*if (ImGui::SliderFloat("Thumbnail Size", &thumbnailSize, 16, 512))
 			Commands::ExecuteRawValueCommand(&thumbnailSize, size, std::string("ContentBrowserPanel-Thumbnail Size"));
 
 		if (ImGui::SliderFloat("Padding", &padding, 0, 32))
-			Commands::ExecuteRawValueCommand(&padding, offset, std::string("ContentBrowserPanel-Padding"));
+			Commands::ExecuteRawValueCommand(&padding, offset, std::string("ContentBrowserPanel-Padding"));*/
```
Comment (szymix200006) [inline Engine-Editor/src/Panels/ContentBrowserPanel.cpp]: Why it's commented, should it be deleted?

File: Engine/premake5.lua [hunk 10]
```diff
@@ -44,6 +44,7 @@ project "Engine"
 		"%{IncludeDir.ImGuizmo}",
```
Comment (szymix200006) [inline Engine/premake5.lua]: Delete this file

File: Engine/src/Engine/Assistant/AssistantManager.h [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+
+		bool ThreadAvailable(std::string threadID);
+
+	private:
+		std::string m_AssistantName;
+		std::string m_AssistantInstructions;
+		std::string m_AssistantID;
+		std::unordered_map<std::string, Thread*> m_Threads;
+		bool newVoiceMessageAvailable;
+		std::string lastVoiceMessage;
```
Comment (szymix200006) [inline Engine/src/Engine/Assistant/AssistantManager.h]: Move the variables with the biggest stride to the front

**Review by szymix200006 — CHANGES_REQUESTED** [review]

File: Engine/src/Engine/Assistant/AssistantManager.h [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+		PyObject* m_InitiateRun;
+		PyObject* m_WaitForCompletion;
+		PyObject* m_GetLastMessage;
+		PyObject* m_CheckIfAssistantExists;
+		PyObject* m_GetToolCall;
+		PyObject* m_SubmitToolOutputs;
+		PyObject* m_InitVoiceAssistant;
+		PyObject* m_CheckMicrophoneAvailable;
+		PyObject* m_SetVolume;
+		PyObject* m_CheckAPI;
```
Comment (szymix200006) [inline Engine/src/Engine/Assistant/AssistantManager.h]: Set member variables to nullptr

File: Engine-Editor/src/Panels/AssistantPanel.h [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+        ImTextureID m_IconReadMessageAloud;
+        bool m_isListening;
+        bool m_shouldListen;
+        bool m_isMicrophoneAvailable;
+        bool m_isLastMessageFromUser;
+        bool m_readAloud;
+        bool m_showMessageTooltip;
+        bool m_isAssistantMessageInProgress;
+        bool m_assistantInitialized;
+        ImGui::MarkdownConfig mdConfig;
```
Comment (szymix200006) [inline Engine-Editor/src/Panels/AssistantPanel.h]: Put the variables with biggest stride first

File: Engine-Editor/src/Panels/AddResourcePanel.h [hunk 14]
```diff
@@ -18,10 +19,25 @@ namespace eg {
...
 		Ref<ImagePanel> m_ImagePanel;
 		Ref<AnimationPanel> m_AnimationPanel;
+		Ref<Texture2D> m_AnimationIcon;
+		Ref<Texture2D> m_ShaderIcon;
+		Ref<Texture2D> m_FontIcon;
+		Ref<Texture2D> m_TextIcon;
+		Ref<Texture2D> m_ImageIcon;
+		Ref<Texture2D> m_ScriptIcon;
+		Ref<Texture2D> m_NativeScriptIcon;
+		Ref<Texture2D> m_CustomIcon;
```
Comment (szymix200006) [inline Engine-Editor/src/Panels/AddResourcePanel.h]: Set refs to nullptr

File: Engine-Editor/src/EditorActions.h [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+	public:
+		static std::string ExecuteEntityAction(const std::string& actionName, std::vector<std::string> params);
+		static std::string ExecuteComponentAction(const std::string& actionName, std::vector<std::string> params);
+
+		static void SetScene(const Ref<Scene>& scene) { m_Scene = scene; }
+		static void SetSelectionContext(Entity* entity) { m_SelectedEntity = entity; }
+
+	private:
+		static Ref<Scene> m_Scene;
+		static Entity* m_SelectedEntity;
```
Comment (szymix200006) [inline Engine-Editor/src/EditorActions.h:68]: Set to nullptr

File: Engine-Editor/src/Panels/AssistantPanel.h [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+        Ref<Texture2D> m_IconCopy;
+        Ref<Texture2D> m_IconSend;
+        Ref<Texture2D> m_IconMicrophone;
+        Ref<Texture2D> m_IconMicrophoneOff;
+        Ref<Texture2D> m_IconMicrophoneUnavailable;
+        Ref<Texture2D> m_IconSettings;
+        Ref<Texture2D> m_IconReadAloud;
+        Ref<Texture2D> m_IconReadAloudHover;
+        Ref<Texture2D> m_IconReadAloudActive;
+        ImTextureID m_IconReadMessageAloud;
```
Comment (szymix200006) [inline Engine-Editor/src/Panels/AssistantPanel.h]: Set to nullptr

**Review by szymix200006 — CHANGES_REQUESTED** [review]

File: Engine-Editor/src/EditorActions.h [hunk 17]
```diff
@@ -7,11 +7,61 @@ namespace eg
 	class EditorActions
 	{
 	public:
-		static std::string ExecuteEntityAction(const std::string& actionName, std::vector<std::string> params);
-		static std::string ExecuteComponentAction(const std::string& actionName, std::vector<std::string> params);
+		static std::string ExecuteEntityAction(const std::string& actionName, std::vector<std::string>& params);
+		static std::string ExecuteComponentAction(const std::string& actionName, std::vector<std::string>& params);
```
Comment (miciex) [inline Engine-Editor/src/EditorActions.h]: vectors should be a const if the params arent set inside the function, which would be weird if it was the case

**Review by miciex — APPROVED** [review]

**Review by szymix200006 — APPROVED** [review]

**Pull request merged**